### DISCUSSION TITLE
ReferenceError for 'crypto' package after upgrading to remix-auth-github@1.8.0

### DISCUSSION THREAD

**Discussion by caprica** [sergiodxa/remix-auth-github#39] [Q&A]
Using remix-auth-github@1.7.0 (and earlier), everything is working.

With no code changes, and only upgrading to remix-auth-github@1.8.0 (with no other package updates), the exact same code fails with a `ReferenceError` when invoking authenticate() inside an action.

```
ReferenceError: crypto is not defined
    at Object.state (file:///xxxxxxxx/node_modules/remix-auth-github/node_modules/remix-auth-oauth2/build/lib/generator.js:18:9)
    at GitHubStrategy.authenticate (file:///xxxxxxxx/node_modules/remix-auth-github/node_modules/remix-auth-oauth2/build/index.js:36:35)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

Anyone seen this and has a solution?

**Comment by sergiodxa** [comment]
Yes the issue is the upgrade of the OAuth2Strategy, the latest release should have been a major one, my mistake, I will do a new major

**Comment by caprica** [comment]
I made some progress but am still having problems.

First, I had to change the runtime Node version from v18 to v20 to get the crypto package.

Second, I had to change the strategy options with some renamed keys, from:

```
    new GitHubStrategy(
        {
            clientID: process.env.GITHUB_CLIENT_ID,
            clientSecret: process.env.GITHUB_CLIENT_SECRET,
            callbackURL: 'http://localhost:3000/api/auth/callback',
        },
        async ({ profile }) => {
            // ...
        },
    ),
```

To:

```
    new GitHubStrategy(
        {
            clientId: process.env.GITHUB_CLIENT_ID, // clientID -> clientId
            clientSecret: process.env.GITHUB_CLIENT_SECRET,
            redirectURI: 'http://localhost:3000/api/auth/callback', // callbackURL -> redirectURI
        },
        async ({ profile }) => {
            // ...
        },
    ),
```

This gets me a little further, but then I hit an error:

```
Missing or invalid 'expires_in' field
```

So far I can't get any further than this.

**Comment by caprica** [comment]
In remix-auth-oauth2/lib/token.js:

```javascript
class Result extends TokenRequestResult {
    toJSON() {
        return {
            ...this.body,
            access_token: this.accessToken(),
            token_type: this.tokenType(),
            expires_in: this.accessTokenExpiresInSeconds(),   // <---
            scope: this.scopes().join(" "),
            refresh_token: this.refreshToken(),
        };
    }
}
```

this.body:

```javascript
{access_token: 'gho_XXXXXXXXXXXXXXXXXX', token_type: 'bearer', scope: 'user:email'}
```

And in oslojs/oauth2:

```
public accessTokenExpiresInSeconds(): number {
	if ("expires_in" in this.body && typeof this.body.expires_in === "number") {
		return this.body.expires_in;
	}
	throw new Error("Missing or invalid 'expires_in' field");
}
```

So `expires_in` property is missing from the response that comes from GitHub?

**Comment by caprica** [comment]
I see also the subsequent code tries to get a refresh token, which also does not exist in this response (I am only using GitHub auth as a user login, I am not using access tokens or refresh tokens to access GitHub resources):

oslojs/oauth2:

```javascript
public refreshToken(): string {
	if ("refresh_token" in this.body && typeof this.body.refresh_token === "string") {
		return this.body.refresh_token;
	}
	throw new Error("Missing or invalid 'refresh_token' field");
}
```

So this too would fail.

**Comment by caprica** [comment]
@sergiodxa sorry to ping you directly, but an upgrade to 2.0 still isn't possible because of this missing or invalid token error I described above. Could you please suggest what might be wrong?

**Comment by caprica** [comment]
Oh of course just after I posted I spotted someone raised this: https://github.com/sergiodxa/remix-auth-github/issues/41

**Comment by m-muiznieks** [comment]
Yes, the problem is real.

**Comment by tsmith123** [comment]
Yeah 2.0 doesn't work for me. Have to use 1.7.0 at the moment which works fine.

**Comment by sohelzerdoumi** [comment]
You can temporarily fix this issue by manually adding the latest remix-auth-oauth2 version in your package.json. It worked for me.

```json
"require": {
   "remix-auth-oauth2": "https://github.com/sergiodxa/remix-auth-oauth2#main",
}
```

remix-auth-oauth2 has fixed it.

https://github.com/sergiodxa/remix-auth-oauth2/blob/b331c4715f1ab9024181e8fccd281d6e5ca80356/src/lib/token.ts#L82
